File: android/PositiveOnlySocial/app/src/main/java/com/example/positiveonlysocial/ui/auth/AuthRequirements.kt
package com.example.positiveonlysocial.ui.auth

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp

/**
 * Client-side mirrors of the backend validation patterns in
 * backend/user_system/constants.py. Keeping these in one place lets the live
 * requirement hints and the form-validity checks share a single source of
 * truth so they can never drift apart.
 *
 *   password     = ^(?=.*[0-9])(?=.*[a-z])(?=.*[A-Z])(?=\S+$).{8,}$
 *   username     = ^\w{10,150}$   (150 is the username column's max_length)
 */
/**
 * A single labelled validation rule. [optional] suggestions don't gate form
 * validity (see [AuthRequirements.allMet]) and render in a neutral state rather
 * than as a pass/fail requirement.
 */
data class Requirement(
    val label: String,
    val didMeetRequirement: Boolean,
    val optional: Boolean = false,
)

object AuthRequirements {

    /**
     * Mirrors MIN_USERNAME_LENGTH / MAX_USERNAME_LENGTH in
     * backend/user_system/constants.py. The maximum is the username column's
     * max_length, so a longer name would fail at the database.
     */
    const val MIN_USERNAME_LENGTH = 10
    const val MAX_USERNAME_LENGTH = 150
    val USERNAME_LENGTH_RANGE = MIN_USERNAME_LENGTH..MAX_USERNAME_LENGTH

    fun password(password: String): List<Requirement> = listOf(
        Requirement("At least 8 characters", password.length >= 8),
        Requirement("At least one number", password.any { it.isDigit() }),
        Requirement("At least one lowercase letter", password.any { it.isLowerCase() }),
        Requirement("At least one uppercase letter", password.any { it.isUpperCase() }),
        // Any non-alphanumeric character counts (the backend accepts them all).
        // isLetterOrDigit is Unicode-aware so accented letters aren't flagged.
        Requirement(
            "Adding special characters (like ! @ # \$ % ^ & * - _) is suggested",
            password.any { !it.isLetterOrDigit() && !it.isWhitespace() },
            optional = true,
        ),
        Requirement("No spaces", password.isNotEmpty() && password.none { it.isWhitespace() }),
    )

    fun username(username: String): List<Requirement> = listOf(
        // Code points, like Python's len() — not UTF-16 units.
        Requirement(
            "Between $MIN_USERNAME_LENGTH and $MAX_USERNAME_LENGTH characters",
            username.codePointCount(0, username.length) in USERNAME_LENGTH_RANGE,
        ),
        Requirement(
            "Letters, numbers, and underscores only",
            username.isNotEmpty() && username.codePoints().allMatch { isWordCodePoint(it) },
        ),
    )

    /**
     * Whether [codePoint] is a word character by Python's `\w` — the backend's
     * username rule: `str.isalnum()` plus `_`, which also admits the
     * letter-number and other-number categories (`Ⅻ`, `①`). Checked per code
     * point, not per `Char`: a supplementary-plane letter is two surrogate
     * `Char`s, neither of which is a letter on its own.
     */
    fun isWordCodePoint(codePoint: Int): Boolean =
        codePoint == '_'.code ||
            Character.isLetterOrDigit(codePoint) ||
            Character.getType(codePoint).let {
                it == Character.LETTER_NUMBER.toInt() || it == Character.OTHER_NUMBER.toInt()
            }

    // Optional suggestions are advisory only and never block submission.
    fun allMet(requirements: List<Requirement>): Boolean =
        requirements.filterNot { it.optional }.all { it.didMeetRequirement }
}

/**
 * Renders a checklist of validation requirements. Required rows show a met/unmet
 * state with color + a ✓/✗ glyph and a per-row content description. Optional
 * suggestions never render as "failed": until satisfied they sit in a neutral
 * state (• / "optional"), switching to met once present.
 */
@Composable
fun RequirementHints(requirements: List<Requirement>) {
    Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
        requirements.forEach { requirement ->
            val glyph = when {
                requirement.didMeetRequirement -> "✓"
                requirement.optional -> "•"
                else -> "✗"
            }
            val status = when {
                requirement.didMeetRequirement -> "met"
                requirement.optional -> "optional"
                else -> "not met"
            }
            Text(
                text = "$glyph ${requirement.label}",
                color = if (requirement.didMeetRequirement) Color(0xFF4CAF50)
                        else MaterialTheme.colorScheme.onSurface.copy(alpha = 0.4f),
                style = MaterialTheme.typography.bodySmall,
                modifier = Modifier.semantics {
                    contentDescription = "${requirement.label}: $status"
                },
            )
        }
    }
}

File: ios/Positive Only Social/Positive Only Social/VibesHelpers/AuthRequirements.swift
//
//  AuthRequirements.swift
//  Vibes
//
//  Client-side mirrors of the backend validation patterns in
//  backend/user_system/constants.py. Keeping these in one place lets the live
//  requirement hints and the form-validity checks share a single source of
//  truth so they can never drift apart.
//
//    password     = ^(?=.*[0-9])(?=.*[a-z])(?=.*[A-Z])(?=\S+$).{8,}$
//    username     = ^\w{10,150}$   (150 is the username column's max_length)
//

import Foundation
import SwiftUI

enum AuthRequirements {
    /// A single labelled validation rule and whether the current input satisfies it.
    struct Requirement: Identifiable {
        let label: String
        let didMeetRequirement: Bool
        /// Optional suggestions don't gate form validity (see allMet) and render in a
        /// neutral state rather than as a pass/fail requirement.
        var optional: Bool = false
        var id: String { label }
    }

    private static func matches(_ value: String, _ pattern: String) -> Bool {
        value.range(of: pattern, options: .regularExpression) != nil
    }

    static func password(_ pwd: String) -> [Requirement] {
        [
            Requirement(label: "At least 8 characters", didMeetRequirement: pwd.count >= 8),
            Requirement(label: "At least one number", didMeetRequirement: matches(pwd, "[0-9]")),
            Requirement(label: "At least one lowercase letter", didMeetRequirement: matches(pwd, "[a-z]")),
            Requirement(label: "At least one uppercase letter", didMeetRequirement: matches(pwd, "[A-Z]")),
            // Any non-alphanumeric character counts (the backend accepts them all).
            // Unicode-aware so it doesn't flag accented letters as "special".
            Requirement(label: "Adding special characters (like ! @ # $ % ^ & * - _) is suggested", didMeetRequirement: matches(pwd, "[^\\p{L}\\p{N}\\s]"), optional: true),
            Requirement(label: "No spaces", didMeetRequirement: !pwd.isEmpty && !matches(pwd, "\\s")),
        ]
    }

    /// Mirrors MIN_USERNAME_LENGTH / MAX_USERNAME_LENGTH in
    /// backend/user_system/constants.py. The maximum is the username column's
    /// max_length, so a longer name would fail at the database.
    static let usernameLengthRange = 10...150

    static func username(_ name: String) -> [Requirement] {
        // Counted in scalars, like Python's len() — not grapheme clusters.
        let length = name.unicodeScalars.count
        return [
            Requirement(
                label: "Between \(usernameLengthRange.lowerBound) and \(usernameLengthRange.upperBound) characters",
                didMeetRequirement: usernameLengthRange.contains(length)),
            Requirement(
                label: "Letters, numbers, and underscores only",
                didMeetRequirement: !name.isEmpty && name.unicodeScalars.allSatisfy(isWordScalar)),
        ]
    }

    /// Whether `scalar` is a word character by Python's `\w` — the backend's
    /// username rule: letters and numbers (`str.isalnum()`) plus `_`. Checked
    /// per scalar by general category rather than with Foundation's `\w`, whose
    /// ICU word set also admits combining marks and connector punctuation that
    /// the backend rejects (`e` + U+0301 would pass here and fail registration).
    static func isWordScalar(_ scalar: Unicode.Scalar) -> Bool {
        if scalar == "_" { return true }
        switch scalar.properties.generalCategory {
        case .uppercaseLetter, .lowercaseLetter, .titlecaseLetter, .modifierLetter, .otherLetter,
             .decimalNumber, .letterNumber, .otherNumber:
            return true
        default:
            return false
        }
    }

    static func allMet(_ requirements: [Requirement]) -> Bool {
        // Optional suggestions are advisory only and never block submission.
        requirements.filter { !$0.optional }.allSatisfy { $0.didMeetRequirement }
    }
}

/// Renders a checklist of validation requirements. Required rows show a met/unmet
/// state with color + an SF Symbol and a per-row accessibility label. Optional
/// suggestions never render as "failed": until satisfied they sit in a neutral
/// state announced as "optional", switching to "met" once present. Shared across
/// the auth screens so the labels, colors, and accessibility behavior can't drift.
struct RequirementHints: View {
    let requirements: [AuthRequirements.Requirement]

    private func symbol(for requirement: AuthRequirements.Requirement) -> String {
        if requirement.didMeetRequirement { return "checkmark.circle.fill" }
        return requirement.optional ? "circle" : "xmark.circle"
    }

    private func status(for requirement: AuthRequirements.Requirement) -> String {
        if requirement.didMeetRequirement { return "met" }
        return requirement.optional ? "optional" : "not met"
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 3) {
            ForEach(requirements) { requirement in
                Label(
                    requirement.label,
                    systemImage: symbol(for: requirement)
                )
                .foregroundColor(requirement.didMeetRequirement ? .green : .secondary)
                .font(.caption)
                .accessibilityLabel("\(requirement.label): \(status(for: requirement))")
            }
        }
    }
}
